### PR TITLE
CAMEL-11571: Added Google Cloud BigQuery component

### PR DESCRIPTION
**Pull request by sarel2**
Added a new component for Goolge Cloud BigQuery with Producer only support

### CODE REVIEW

**Comment by oscerd** [comment]
I'll review soon, thanks for the moment :-)

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+ */
+@UriEndpoint(scheme = "bigquery", title = "BigQuery", syntax = "bigquery:projectId:datasetId[:tableName]", label = "messaging")
+public class GoogleBigQueryEndpoint extends DefaultEndpoint {
+    @UriParam
+    protected final GoogleBigQueryConfiguration configuration;
+
+    private final Bigquery bigquery;
+    private final String uri;
+
+    private final Map<String, Boolean> verifiedTables = new ConcurrentHashMap<>();
```
Comment (aldettinger) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java]: `verifiedTables` seems to be unused. Couldn't we remove it ?

File: components/camel-google-bigquery/src/test/java/org/apache/camel/component/google/bigquery/integration/AsyncTest.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    protected RouteBuilder createRouteBuilder() throws Exception {
+        return new RouteBuilder() {
+            public void configure() {
+                from(directIn)
+                        .to("seda:seda");
+                from("seda:seda")
+                        .routeId("Async")
+                        //.threads(10)
```
Comment (aldettinger) [inline components/camel-google-bigquery/src/test/java/org/apache/camel/component/google/bigquery/integration/AsyncTest.java]: We may remove this line too.

**Comment by oscerd** [comment]
Thanks

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ */
+@UriEndpoint(scheme = "bigquery", title = "BigQuery", syntax = "bigquery:projectId:datasetId[:tableName]", label = "messaging")
+public class GoogleBigQueryEndpoint extends DefaultEndpoint {
+    @UriParam
+    protected final GoogleBigQueryConfiguration configuration;
+
+    private final Bigquery bigquery;
+    private final String uri;
+
+    private ExecutorService executorService;
```
Comment (lburgazzoli) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java]: I don't see this executor service being stopped/shutdown at any point

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    @Override
+    public GoogleBigQueryEndpoint getEndpoint() {
+        return (GoogleBigQueryEndpoint) super.getEndpoint();
+    }
+
+    @Override
+    public boolean process(Exchange exchange, AsyncCallback callback) {
+        getEndpoint().getExecutorService().submit(() -> {
```
Comment (lburgazzoli) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java]: As the executor is only used here, maybe the Processor could take executor as constructor argument.

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+     *     <li>A map where all map keys will map to field records. One map object maps to one bigquery row</li>
+     *     <li>A list of maps. Each entry in the list will map to one bigquery row</li>
+     * </ul>
+     * The incoming message is expected to be a List of Maps
+     * The assumptions:
+     * - All incoming records go into the same table
+     * - Incoming records sorted by the timestamp
+     */
+    @Override
+    public void process(Exchange exchange) throws Exception {
```
Comment (lburgazzoli) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java]: ``DefaultAsyncProducer`` by default [wraps](https://github.com/apache/camel/blob/master/camel-core/src/main/java/org/apache/camel/impl/DefaultAsyncProducer.java#L36-L38) this ``process(Exchange)`` method and calls the async one, any reason to do the opposite here ?

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        return producer;
+    }
+
+
+    public Consumer createConsumer(Processor processor) throws Exception {
+        throw new UnsupportedOperationException("Cannot consume from the BigQuery endpoint: " + getEndpointUri());
+    }
+
+    public boolean isSingleton() {
```
Comment (davsclaus) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java]: Avoid non singleton at all cost. Instead you should store the local state on the producer/consumers, such as its executor service etc. The endpoint should generally just be a stateless class.

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        return bigquery;
+    }
+
+    public ExecutorService getExecutorService() {
+        return executorService;
+    }
+
+    @Override
+    protected String createEndpointUri() {
+        return uri;
```
Comment (davsclaus) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java]: You should usually not override this method

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+    private final Logger logger;
+    private final GoogleBigQueryConfiguration configuration;
+
+    public GoogleBigQueryProducer(GoogleBigQueryEndpoint endpoint, GoogleBigQueryConfiguration configuration) {
+        super(endpoint);
+
+        this.configuration = configuration;
+
+        String loggerId = configuration.getLoggerId();
```
Comment (davsclaus) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java]: Why do you want to allow setting logger this way. Please make it just like all the others as a static logger.

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java (hunk 5 above)
Comment (davsclaus) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java]: What is the point with creating a list<exchanges> in a producer??? A producer should generally just store the incoming single exchange.

File: components/camel-google-bigquery/src/test/java/org/apache/camel/component/google/bigquery/unit/GoogleBigQueryProducerTest.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    @Test
+    public void groupExchangeDifferentTableSuffix() throws Exception {
+        Exchange exchange1 = createExchangeWithBody(new HashMap<>());
+        exchange1.getIn().setHeader(GoogleBigQueryConstants.TABLE_SUFFIX, "_SUFFIX1");
+        Exchange exchange2 = createExchangeWithBody(new HashMap<>());
+        exchange2.getIn().setHeader(GoogleBigQueryConstants.TABLE_SUFFIX, "_SUFFIX2");
+        Exchange exchange = createExchangeWithBody(null);
+        exchange.setProperty(Exchange.GROUPED_EXCHANGE, Lists.newArrayList(exchange1, exchange2));
```
Comment (davsclaus) [inline components/camel-google-bigquery/src/test/java/org/apache/camel/component/google/bigquery/unit/GoogleBigQueryProducerTest.java]: Avoid using Exchange.GROUPED_EXCHANGE as its not for components to use, its for the aggregate EIP pattern.

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java (hunk 6 above)
Comment (sarelp) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryEndpoint.java]: Changed to true

File: components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java (hunk 5 above)
Comment (sarelp) [inline components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java]: Changed to DefaultProducer removing need for executor. Rest client is synchronous. Will need to revisit when changing to async rpc client in future

**Comment by sarelp** [comment]
@oscerd thanks have added missing descriptors.

**Comment by oscerd** [comment]
I'll take care of adding the starter too in the descriptors. I'm gonna merge today.

**Comment by oscerd** [comment]
Thanks @sarelp the PR has been merged. Can you close this?

**Pull request closed without merging**